core/src/main/scala/org/apache/spark/sql/TiSparkFunctionRegistry.scala

```diff
@@ -0,0 +1,114 @@
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.analysis.FunctionRegistry.FunctionBuilder
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, ExpressionInfo, Microsecond, RuntimeReplaceable, TimeToSec, ToDays, ToSeconds, WeekDay, YearWeek}
+
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
+
+object TiSparkFunctionRegistry {
+
+// Note: Whenever we add a new entry here, make sure we also update ExpressionToSQLSuite
+  val expressions: Map[String, (ExpressionInfo, FunctionBuilder)] = Map(
+    // misc non-aggregate functions
+    expression[Microsecond]("microsecond"),
+    expression[WeekDay]("weekday"),
+    expression[ToSeconds]("to_seconds"),
+    expression[ToDays]("to_days"),
+    expression[TimeToSec]("time_to_sec"),
+    expression[YearWeek]("yearweek")
+    // TODO: this need import antler 4 into our code base.
+    // the behavior is like the following:
+    // select extract year from '2019-01-10' which is just
+    // identical to select year('2019-01-10')
+    // expression[Extract]("extract")
+  )
+
+  /** See usage above. */
+  private def expression[T <: Expression](
+    name: String
+  )(implicit tag: ClassTag[T]): (String, (ExpressionInfo, FunctionBuilder)) = {
+
+    // For `RuntimeReplaceable`, skip the constructor with most arguments, which is the main
+    // constructor and contains non-parameter `child` and should not be used as function builder.
+    val constructors = if (classOf[RuntimeReplaceable].isAssignableFrom(tag.runtimeClass)) {
```

Comment:
If this logic comes from Spark, please add comments refers to spark. When we upgrade, we need to check those logic in case bug fix ignored.